File: electives/JPP/L3/cpp/test/test_User.cpp
#include <random> 
#include <catch2/catch_test_macros.hpp> 
#include <catch2/generators/catch_generators_adapters.hpp>
#include <catch2/generators/catch_generators_random.hpp>
#include "User.hpp"
#include "GF.hpp"

TEST_CASE("User: p = 1234567891"){
    constexpr uint32_t p = 1234567891; 
    constexpr uint32_t m = 20;
    GF message = GF(p, m);

    DHSetup<GF> DH = DHSetup<GF>(p);
    User<GF> user1 = User<GF>(&DH);
    User<GF> user2 = User<GF>(&DH);

    REQUIRE(user1.getPublicKey() != user2.getPublicKey()); 
    REQUIRE_THROWS(user1.encrypt(message), user2.encrypt(message),
                    user1.decrypt(message), user2.decrypt(message));

    user1.setKey(user2.getPublicKey());
    user2.setKey(user1.getPublicKey());
    
    GF encryptedMessage = user1.encrypt(message);
    GF decryptedMessage = user2.decrypt(encryptedMessage);
    
    REQUIRE(decryptedMessage == message);
};

File: electives/JPP/L3/python/test_User.py
import unittest
from User import User
from DHSetup import DHSetup
from GF import GF

class TestUser(unittest.TestCase):
    def test(self):
        p = 1234567891
        m = 20
        message = GF(p, m)
        DH = DHSetup(p)
        user1 = User(DH)
        user2 = User(DH)

        self.assertNotEqual(user1.get_public_key(), user2.get_public_key())

        user1.set_key(user2.get_public_key())
        user2.set_key(user1.get_public_key())

        encryptedMessage = user1.encrypt(message)
        decryptedMessage = user2.decrypt(encryptedMessage)

        self.assertEqual(decryptedMessage.get_val(), message.get_val())

if __name__ == '__main__':
    unittest.main()
